rclpy/rclpy/node.py

```diff
@@ -153,6 +153,76 @@ def __init__(
         if start_parameter_services:
             self._parameter_service = ParameterService(self)
 
+    @property
+    def publishers(self) -> List[Publisher]:
+        """
+        Return publishers that have been created on this node.
+
+        .. warning::
+        The returned list must not be modified.
```

Comment:
The API could provide a tuple or copied list instead?

Reply:
That means the executor would copy lists every time it rebuilds the wait set.

https://github.com/ros2/rclpy/blob/d672075cd5a821314af2e460af68b90373c5506f/rclpy/rclpy/executors.py#L455-L460

I don't know if that's a big deal; I could profile it.

Reply:
Was just an idea to prevent callers from modifying the lists...

Reply:
Actually, this could `yield from` and be a generator to avoid copying stuff

Reply:
236b20a makes these generators. This is definitely an API change now though.

Reply:
Interesting idea. I don't think I have seen this combination. :+1: